File: backend_server/src/middleware/correlationId.js
const crypto = require('crypto');

/**
 * Correlation ID middleware.
 * Reads X-Request-ID from the incoming request (set by the client or a
 * load balancer) or generates a new UUID. Attaches the ID to req.requestId
 * and echoes it back in the X-Request-ID response header so callers can
 * link their log entries to backend log entries.
 */
const correlationId = (req, res, next) => {
    const id = req.headers['x-request-id'] || crypto.randomUUID();
    req.requestId = id;
    res.setHeader('X-Request-ID', id);
    next();
};

module.exports = correlationId;

File: ai_server/src/middleware/correlation_id.py
"""
Correlation ID middleware for request tracing.

Reads X-Request-ID from the incoming request (forwarded by the backend) or
generates a new UUID. Attaches the ID to the response header so callers can
trace a request end-to-end across backend and AI server logs.

The request_id ContextVar allows route handlers and services to include the
ID in their own log lines without it being passed as an argument.
"""
import uuid
from contextvars import ContextVar
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

# Per-request context variable — set at middleware entry, readable anywhere
# in the same async call chain via request_id_var.get()
request_id_var: ContextVar[str] = ContextVar("request_id", default="")


class CorrelationIdMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next) -> Response:
        req_id = request.headers.get("x-request-id") or str(uuid.uuid4())
        token = request_id_var.set(req_id)
        try:
            response = await call_next(request)
        finally:
            request_id_var.reset(token)
        response.headers["X-Request-ID"] = req_id
        return response
